packages/lib/src/__tests__/login-authenticator.test.ts

```diff
@@ -0,0 +1,475 @@
+	const profileName = 'myProfile'
+	const clientIdProvider = {
+		baseURL: 'https://example.com/unused-here',
+		authURL: 'https://example.com/unused-here',
+		keyApiURL: 'https://example.com/unused-here',
+		baseOAuthInURL: 'https://example.com/oauth-in-url',
+		oauthAuthTokenRefreshURL: 'https://example.com/refresh-url',
+		clientId: 'client-id',
+	}
+	const config = {
+		appenders: {
+			main: { type: 'recording' },
+			stdout: { type: 'stdout' },
+		},
+		categories: {
+			default: { appenders: ['main'], level: 'ERROR' },
+			'login-authenticator': { appenders: ['main'], level: 'TRACE' },
+		},
+	}
+	const credentialsFileData = {
+		[profileName]: {
+			'accessToken': 'an access token',
+			'refreshToken': 'a refresh token',
+			'expires': '2020-10-15T13:26:39.966Z',
+			'scope': 'controller:stCli',
+			'installedAppId': 'installed app id',
+			'deviceId': 'device id',
+		},
+	}
+	interface AuthTokenResponse {
+		access_token: string
+		refresh_token: string
+		expires_in: number
+		scope: string
+		installed_app_id: string
+		device_id: string
+	}
+	const tokenResponse: AxiosResponse<AuthTokenResponse> = {
+		data: {
+			access_token: 'access token',
+			refresh_token: 'refresh token',
+			expires_in: 24 * 3600,
+			scope: 'scope list',
+			installed_app_id: 'installed app id',
+			device_id: 'device id',
+		},
+		status: 200,
+		statusText: 'OK',
+		headers: '',
+		config: {},
+	}
+
+	const mockServer = {
+		close: jest.fn(),
+	}
+
+	mockApp.listen.mockReturnValue(mockServer)
+
+	const mockStartResponse = {
+		redirect: jest.fn(),
+	}
+	const mockFinishRequest = {
+		query: { code: 'auth-code' },
+	}
+	const mockFinishResponse = {
+		send: jest.fn(),
+	}
+
+
+	logManager.init(config)
+
+	const mkdirMock = fs.mkdirSync as jest.Mock<typeof fs.mkdirSync>
+	const readFileMock = (fs.readFileSync as unknown as jest.Mock<Buffer, [fs.PathLike]>)
+		.mockImplementation(() => { throw { code: 'ENOENT' } })
+	const writeFileMock = fs.writeFileSync as jest.Mock<typeof fs.writeFileSync>
+	const chmodMock = fs.chmod as unknown as jest.Mock<void, [PathLike, string | number, NoParamCallback]>
+
+	const getPortMock = (getPort as unknown as jest.Mock<Promise<number>, [getPort.Options | undefined]>)
+		.mockResolvedValue(7777)
+	const postMock = (axios.post as unknown as jest.Mock<Promise<AxiosResponse<unknown>>, [string, unknown, AxiosRequestConfig | undefined]>)
+		.mockResolvedValue(tokenResponse)
+
+	function setupAuthenticator(): LoginAuthenticator {
+		LoginAuthenticator.init(credentialsFilename)
+		return new LoginAuthenticator(profileName, clientIdProvider)
+	}
+
+	type ExpressRouteHandler = (request: Request, response: Response) => void
+	const finishHappy = (finishHandler: ExpressRouteHandler): void =>
+		finishHandler(mockFinishRequest as unknown as Request, mockFinishResponse as unknown as Response)
+	// Call handlers like would happen if the user were following the flow in the browser.
```

Comment:
I think this feels too much like an integration test rather than unit tests with the delays in here. I see a couple options to remove the need for this imitation and still get our code tested.

1. Extract the arrow functions out of `login()` so that they can be tested separately. Then login tests can just assert that the correct functions were called the expected times.
2. Alternatively, we could start a separate test collection for "integration" and add the tests using `imitateBrowser` to that so that there aren't delays in the unit tests. CI could run them separately as well.

Reply:
The delays are only to let the promise finish...they are very quick. They are unit tests, not integration tests...they don't test anything outside this class. The `imitateBrowser` method is what allows them to be unit tests...it's just calling the methods that would get called via the browser if it were an integration test or running for real.

Reply:
Actually not to let the promise finish...an `await` would do that. I needed to let enough of the promise finish to be able to get access to `express.get` calls so I could feign what the browser would do before the whole promise could finish.

Reply:
Yes since the express server is mocked, it is a unit test. I just thought the name `imitateBrowser` along with the requirement to delay felt off.

Generally I don't think static delays in our unit tests are a good idea even if they are quick. We can't really guarantee that something won't change to slightly increase the required time and cause flakey behavior. I will approve since this works and gets this under test, but still recommend a refactor to remove the delays.